### ISSUE TITLE
(WIP) UTXO pool implementation

### ISSUE THREAD

**Issue by sbellem** [bigchaindb/bigchaindb#1867]
### Problems & Concerns
This proposal wishes to explore ways in which the UTXO could be simplified such that it holds what is strictly essential for the purpose of validating transactions (as far as unspent outputs are concerned), when ABCI primitives such as `check_tx` and `deliver_tx` are invoked.

Some questions that this proposal wishes to address could be roughly formulated as follows:

* How can we keep the UTXO as lightweight as possible?
* How can we make querying the UTXO as fast as possible?

#### `utxo_hash`
Essential information for one UTXO:
* transaction id
* output index
* condition uri
* amount
* asset id

the above 5 pieces of information could be concatenated and passed to a hash function in order to generate a unique hash representing the unspent output (UTXO). The resulting hash (`utxo_hash`) could be used as a key in a key/value store whereas the tuple `(txid, output_index, condition_uri, amount, asset_id)` would be the value.

```python
from sha3 import sha3_256

txid = '5dd80cc5df2547226cc5b62832d719772883115a3d51ce5979c921c866a394cd'
output_index = 0
condition_uri = 'ni:///sha-256;7ErbKLgZNqjESdI7AHIqY9Nr2wkoqyMsFwDVfezk_Cc?fpt=ed25519-sha-256&cost=131072'
amount = '1'
asset_id = '107ec21f4c53cd2a934941010437ac74882161bcbefdfd7664268823fc347996'

utxo_hash = sha3_256('{}{}{}{}{}'.format(
    txid, output_index, condition_uri, amount, asset_id).encode()).hexdigest()
```
```pyhton
>>> utxo_hash
'df82094663c535a1a35b55b92d7a6e331a95a3e5cf78f27f940497a4d60a9e5f'
```

#### `utxo_hashes` --> `utxo_merkle_root`
The utxo hashes could perhaps be stored in a merkle tree and be used to compute a merkle root, which could be used to define the state of the UTXO pool. 

An agreed deterministic order (e.g.: lexicographical) would need to be defined to make sure that different nodes compute the same merkle root from the same UTXO pool. 

This merkle root could be passed as the app hash (#1865).

This should allow for efficient verifications such as proof of membership and proof  of non membership.

**Comment by kansi** [comment]
> An agreed deterministic order (e.g.: lexicographical) would need to be defined to make sure that different nodes compute the same merkle root from the same UTXO pool.

Would treap be helpful in ensuring this?

**Comment by sbellem** [comment]
### Potential problem
Fulfillments are not captured by the transaction id/hash. When a transaction id is created the transaction payload without fulfillments is hashed. See https://docs.bigchaindb.com/projects/py-driver/en/latest/handcraft.html#up-to-now and https://docs.bigchaindb.com/projects/py-driver/en/latest/handcraft.html#id for instance.

**Comment by ttmc** [comment]
If a transaction is assumed to be _valid_ (as all in-chain transactions are, in the Tendermint world), then its transaction ID should match its entire contents, including the `fulfillment` values.

I guess the question is: does the transaction with its `fulfillment` values set to `null` (`fulfillment_message`) uniquely determine the `fulfillment` strings (and therefore the _entire_ contents of the transaction)?

Each `fulfillment` value in a transaction is a deterministic function of 1) the transaction's `fulfillment_message` and 2) some private keys which depend on the condition that being fulfilled, a condition which doesn't actually _live_ in the transaction at all; it lives in the output of a previous (valid) transaction.

Information about that previous condition got propagated to the current transaction inside the `fulfills` object of the input in question, inside the transaction ID where that previous condition lives...

The information flow is getting hard to follow, which raises a big red flag for me: one should try to keep it simple, since hard stuff is hard to maintain and debug.

In #1874, I proposed a breaking change to how the transaction ID is computed:

> One way to change this would be to calculate the transaction ID _last_, once all `fulfillment` strings are known, and with their values (string values) left intact (not replaced with `null`).

(The `fulfillment_message` would have to change to not depend on the transaction ID, but that's okay since it contains no new information anyway.)

If we did that, then it's obvious that the transaction ID would depend on _everything_ inside the transaction (except for the `id` itself, but adding that is impossible due to recursion), and therefore it would act as a good transaction fingerprint.

**Comment by sbellem** [comment]
> If a transaction is assumed to be valid (as all in-chain transactions are, in the Tendermint world), then its transaction ID should match its entire contents, including the fulfillment values.

Yes.

Some thoughts, slightly scattered.

If I am not wrong, one goal of the app hash, is to check whether the validating nodes are in the same state after a block has been committed. Or said differently, at the start of a new round in which a block needs to be agreed on, the validating nodes will also take into account the app hash. I am not sure on how exactly it is used in the consensus part; the code would need to be inspected to have a clear understanding. Perhaps there is some documentation too on that matter.

The point though is that the app hash should be usable to detect diverging states. Now if two nodes each have two identical app hashes, and if the app hash is derived from the transaction hash and output index, then we cannot detect diverging inputs/fulfillments.

Moreover, since the fulfillment is a crucial piece of the validation step, it seems to be a scary situation to think that two nodes diverge on the fulfillment. For instance if one was to replay validation on the chain for one of those two diverging nodes one validation should fail.

Hence, it seems to me that the fulfillments are an essential part of the state. Two nodes are only in the same state if and only if their stored UTXO is totally identical.

It probably makes sense to treat the state question as different from a uniqueness property.

A properly working instance of BigchainDB will reject a transaction with an incorrect fulfillment, and consequently there cannot be two transactions with the same id but with different fulfillments. So in that sense a transaction without its fulfillment(s) is unique, in a node that is following the protocol.

The problem is perhaps not so much with uniqueness but with not detecting a (byzantine) node that would be deviating from the protocol and storing a messed up DAG, as it would ignore the validation rules and would tolerate wrong fulfillments, but yet would still always present a "valid" app hash, aligned with the correct nodes.
